ribbon-eureka/src/main/java/com/netflix/loadbalancer/EurekaZoneAwareLoadBalancer.java

```diff
@@ -0,0 +1,157 @@
+        super.setServerListForZones(zoneServersMap);
+        if (balancers == null) {
+            balancers = new ConcurrentHashMap<String, BaseLoadBalancer>();
+        }
+        for (Map.Entry<String, List<Server>> entry: zoneServersMap.entrySet()) {
+            String zone = entry.getKey().toLowerCase();
+            getLoadBalancer(zone).setServersList(entry.getValue());
+        }
+        // check if there is any zone that no longer has a server
+        // and set the list to empty so that the zone related metrics does not
+        // contain stale data
+        for (Map.Entry<String, BaseLoadBalancer> existingLBEntry: balancers.entrySet()) {
+            if (!zoneServersMap.keySet().contains(existingLBEntry.getKey())) {
+                existingLBEntry.getValue().setServersList(Collections.emptyList());
+            }
+        }
+    }
+
+    @Override
+    public Server chooseServer(Object key) {
+        if (!ZoneAwareLoadBalancerConfig.getEnabled() || getLoadBalancerStats().getAvailableZones().size() <= 1) {
+            logger.debug("Zone aware logic disabled or there is only one zone");
+            return super.chooseServer(key);
+        }
+        Server server = null;
+        try {
+            LoadBalancerStats lbStats = getLoadBalancerStats();
+            Map<String, ZoneSnapshot> zoneSnapshot = ZoneAvoidanceRule.createSnapshot(lbStats);
+            logger.debug("Zone snapshots: {}", zoneSnapshot);
+            if (triggeringLoad == null) {
+                triggeringLoad = ZoneAwareLoadBalancerConfig.newTriggeringLoadProperty(this.getName());
+            }
+
+            if (triggeringBlackoutPercentage == null) {
+                triggeringBlackoutPercentage = ZoneAwareLoadBalancerConfig.newTriggeringBlackoutPercentageProperty(this.getName());
+            }
+
+            Set<String> availableZones = ZoneAvoidanceRule.getAvailableZones(zoneSnapshot, triggeringLoad.get(), triggeringBlackoutPercentage.get());
+            logger.debug("Available zones: {}", availableZones);
+            if (availableZones != null &&  availableZones.size() < zoneSnapshot.keySet().size()) {
+                String zone = ZoneAvoidanceRule.randomChooseZone(zoneSnapshot, availableZones);
+                logger.debug("Zone chosen: {}", zone);
+                if (zone != null) {
+                    BaseLoadBalancer zoneLoadBalancer = getLoadBalancer(zone);
+                    server = zoneLoadBalancer.chooseServer(key);
+                }
+            }
+        } catch (Throwable e) {
+            logger.error("Unexpected exception when choosing server using zone aware logic", e);
+        }
+        if (server != null) {
+            return server;
+        } else {
+            logger.debug("Zone avoidance logic is not invoked.");
+            return super.chooseServer(key);
+        }
+    }
+
+    @VisibleForTesting
+    BaseLoadBalancer getLoadBalancer(String zone) {
+        zone = zone.toLowerCase();
```

Comment:
Just took this from the existing ZoneAwareLoadBalancer. It was pretty much impossible to refactor so that I can use the existing ZoneAwareLoadBalancer which is why some code is duplicated.